torch_np/_wrapper.py

```diff
@@ -310,7 +310,19 @@ def full(shape, fill_value, dtype=None, order="C", *, like=None):
         raise NotImplementedError
     if isinstance(fill_value, ndarray):
         fill_value = fill_value.get()
-    torch_dtype = _dtypes.torch_dtype_from(dtype)
+    if dtype is None:
+        if isinstance(fill_value, bool):
+            torch_dtype = torch.bool
+        elif isinstance(fill_value, int):
+            torch_dtype = torch.int64
+        elif isinstance(fill_value, float):
+            torch_dtype = torch.float64
+        elif isinstance(fill_value, complex):
+            torch_dtype = torch.complex128
+        else:
+            torch_dtype = _dtypes.torch_dtype_from(dtype)
+    else:
+        torch_dtype = _dtypes.torch_dtype_from(dtype)
```

Comment:
NB: I thought the if-elif chain is taken care of already in `torch_dtype_from`:

```
In [3]: tnp._dtypes.torch_dtype_from(complex)
Out[3]: torch.complex128

In [4]: tnp._dtypes.torch_dtype_from(float)
Out[4]: torch.float64

In [5]: tnp._dtypes.torch_dtype_from(bool)
Out[5]: torch.bool

In [6]: tnp._dtypes.torch_dtype_from(int)
Out[6]: torch.int64
```
(am going to move `_dtype_from_torch` wholesale into `@_decorators.dtype_to_torch` decorator, cf `eye`).
That said, this all is a bit in flux, so as long as it's tested, this particular code can be as ugly as needed --- I'm going to take this along in the next refactor pass.

Reply:
`torch_dtype_from(type(fill_value))` wouldn't follow the expected behaviour for dtype inference, e.g.

```python
>>> torch_dtype_from(int)
uint8  # we need int64 for dtype inference in full()
```

Reply:
Yep you're right, I didn't bother to check this :facepalm: Pushed a commit that just does `torch_dtype_from(type(fill_value))`